### PR TITLE
[Feature] StarOS worker update file store

### PR DESCRIPTION
**Pull request by TszKitLo40**
Fixes #issue

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

File: be/src/service/staros_worker.cpp [hunk 1]
```diff
@@ -300,6 +347,19 @@ absl::StatusOr<std::shared_ptr<StarOSWorker::FileSystem>> StarOSWorker::new_shar
     return std::move(fs);
 }
 
+CacheKey StarOSWorker::get_cache_key(std::string_view scheme, const Configuration& conf) {
```
Comment (kevincai) [inline be/src/service/staros_worker.cpp:350]: can be a static function, touches  nothing of StarOSWorker member variable.

File: be/src/service/staros_worker.cpp [hunk 2]
```diff
@@ -58,7 +58,30 @@ StarOSWorker::~StarOSWorker() = default;
 
 absl::Status StarOSWorker::add_shard(const ShardInfo& shard) {
     std::unique_lock l(_mtx);
-    _shards.try_emplace(shard.id, ShardInfoDetails(shard));
+    auto it = _shards.find(shard.id);
+    if (it != _shards.end() &&
+        it->second.shard_info.path_info.fs_info().version() < shard.path_info.fs_info().version()) {
```
Comment (kevincai) [inline be/src/service/staros_worker.cpp:63]: possibly, `has_fs_info()` is false?

**Review by kevincai — COMMENTED** [review]
TBD

File: be/src/service/staros_worker.cpp [hunk 3]
```diff
@@ -159,79 +182,114 @@ absl::StatusOr<std::shared_ptr<fslib::FileSystem>> StarOSWorker::build_filesyste
...
         scheme = "azblob://";
-        {
+        break;
+    default:
+        return absl::InvalidArgumentError("Unknown shard storage scheme!");
+    }
+    return scheme;
+}
+
+absl::Status StarOSWorker::build_conf_from_shard_info(const ShardInfo& info, Configuration& localconf) {
```
Comment (kevincai) [inline be/src/service/staros_worker.cpp:228]:
```suggestion
absl::StatusOr<Configuration> StarOSWorker::build_conf_from_shard_info(const ShardInfo& info) {
```

File: be/src/service/staros_worker.cpp [hunk 4]
```diff
@@ -159,79 +182,114 @@ absl::StatusOr<std::shared_ptr<fslib::FileSystem>> StarOSWorker::build_filesyste
...
+    // FIXME: currently the cache root dir is set from be.conf, could be changed in future
+    std::string cache_dir = config::starlet_cache_dir;
```
Comment (kevincai) [inline be/src/service/staros_worker.cpp:230]: looks like we can extract these lines into a small static function `need_enable_cache(CacheInfo)`

File: be/src/service/staros_worker.cpp [hunk 5]
```diff
@@ -159,113 +181,139 @@ absl::StatusOr<std::shared_ptr<fslib::FileSystem>> StarOSWorker::build_filesyste
...
+    auto scheme = build_scheme_from_shard_info(info);
+    if (!scheme.ok()) {
+        return scheme.status();
+    }
 
-    // FIXME: currently the cache root dir is set from be.conf, could be changed in future
     std::string cache_dir = config::starlet_cache_dir;
     auto cache_info = info.cache_info;
     bool cache_enabled = cache_info.enable_cache() && !cache_dir.empty();
+    if (cache_enabled) {
```
Comment (kevincai) [inline be/src/service/staros_worker.cpp:196]:
```suggestion
    if (need_enable_cache(info)) {
```

File: be/src/service/staros_worker.cpp [hunk 6]
```diff
@@ -159,113 +181,139 @@ absl::StatusOr<std::shared_ptr<fslib::FileSystem>> StarOSWorker::build_filesyste
...
+        // set environ variable to cachefs directory
+        setenv(fslib::kFslibCacheDir.c_str(), cache_dir.c_str(), 0 /*overwrite*/);
+    }
+    return new_shared_filesystem(*scheme, *localconf);
+}
+
+bool StarOSWorker::need_enable_cache(const ShardInfo& info) {
+    // FIXME: currently the cache root dir is set from be.conf, could be changed in future
+    std::string cache_dir = config::starlet_cache_dir;
```
Comment (kevincai) [inline be/src/service/staros_worker.cpp:205]: this var is not necessary any more, use config::starlet_cache_dir directly.

**Review by wyb — APPROVED** [review]

**Review by kevincai — APPROVED** [review]

**Review by sduzh — APPROVED** [review]

**Comment by TszKitLo40** [comment]
https://github.com/Mergifyio backport branch-3.1

**Pull request merged**